portfolio/src/main/java/com/google/sps/data/FindSchedule.java

```diff
@@ -0,0 +1,112 @@
+// the scheduling algorithms can be their own class that extends this one.
+
+public class FindSchedule {
+  private static final Comparator<CalendarEvent> sortByEventStartTimeAscending =
+      Comparator.comparing(CalendarEvent::getStartTimeInstant);
+
+  private static final Comparator<Task> sortByTaskDurationThenName =
+      Comparator.comparing(Task::getDuration).thenComparing(Task::getName);
+
+  /**
+   * This method schedules tasks from shortest to longest and returns a ScheduledTask Collection
+   * based on the tasks that were able to be scheduled.
+   */
+  public static Collection<ScheduledTask> shortestTaskFirst(
+      Collection<CalendarEvent> events,
+      Collection<Task> tasks,
+      Instant workHoursStartTime,
+      Instant workHoursEndTime) {
+    List<CalendarEvent> eventsList = new ArrayList<CalendarEvent>(events);
+    List<Task> tasksList = new ArrayList<Task>(tasks);
+    Collections.sort(eventsList, sortByEventStartTimeAscending);
+    Collections.sort(tasksList, sortByTaskDurationThenName);
+    List<TimeRange> availableTimes =
+        getEmptyTimeRanges(eventsList, workHoursStartTime, workHoursEndTime);
+    List<ScheduledTask> scheduledTasks = new ArrayList<ScheduledTask>();
+    int rangeIndex = 0;
+    int taskIndex = 0;
+    // Instant indicating the start time we are currently trying to schedule
+    // events in.
+    Instant currentScheduleTime = workHoursStartTime;
+    // This will iterate through the time ranges and tasks and if one can be
+    // scheduled then it will be (and we move onto the next task) otherwise
+    // we move on to the next range (this is because the tasks are sorted by
+    // duration so if one task did not fit in the given range then we know no
+    // later ones will fit either). We create new Task objects for the result
+    // so data structures passed in are never changed.
+    while (rangeIndex < availableTimes.size() && taskIndex < tasksList.size()) {
+      TimeRange availableTimeRange = availableTimes.get(rangeIndex);
+      Task task = tasksList.get(taskIndex);
+      // Either time is already past the start of the time range or we should
+      // update it (maybe this is our first iteration in the range).
+      if (availableTimeRange.start().isAfter(currentScheduleTime)) {
+        currentScheduleTime = availableTimeRange.start();
+      }
+      // The task can be scheduled in the current time range.
+      if (!currentScheduleTime
+          .plusSeconds(task.getDuration().getSeconds())
+          .isAfter(availableTimeRange.end())) {
+        ScheduledTask scheduledTask = new ScheduledTask(task, currentScheduleTime);
+        scheduledTasks.add(scheduledTask);
+        currentScheduleTime = currentScheduleTime.plusSeconds(task.getDuration().getSeconds());
+        taskIndex++;
+      } else {
+        rangeIndex++;
+      }
+    }
+
+  /**
+   * This method returns an TimeRange array which represent the periods of time that are empty of
```

Comment:
array => list. Arrays and Lists are different concepts in Java.

Reply:
Done.